source/sigv4.c

```diff
@@ -854,12 +894,44 @@ static SigV4Status_t getCredentialScope( SigV4Parameters_t * pSigV4Params,
     }
 
 /*-----------------------------------------------------------*/
-    static SigV4Status_t writeSignedHeaderToString( size_t headerIndex,
-                                                    uint32_t flags,
-                                                    CanonicalContext_t * canonicalRequest )
+
+    static bool isTrimmableSpace( const char * value,
+                                  size_t index,
+                                  size_t valLen,
+                                  size_t trimmedLength )
+    {
+        bool ret = false;
+
+        /* Only trim spaces. */
+        if( isspace( value[ index ] ) )
+        {
+            /* The last character is a trailing space. */
+            if( ( index + 1 ) == valLen )
+            {
+                ret = true;
+            }
+            /* Trim if the next character is also a space. */
+            else if( isspace( value[ index + 1 ] ) )
+            {
+                ret = true;
+            }
+            /* It is a leading space if no characters have been written yet. */
+            else if( trimmedLength == 0U )
+            {
+                ret = true;
+            }
+        }
+
+        return ret;
```

Comment:
Alternative suggestion to make the code very explicit about the different types of spaces that are trimmed.
```suggestion

    bool isCurrCharSpace = isSpace( value[index] ); 
    bool isLeadingSpace = isCurrCharSpace && ( trimmedLength == 0 );
    bool isTrailingSpace = isCurrCharSpace && ( ( index + 1 ) == valLen ); 
    bool isNotLastInSequentialSpaces = isCurrCharSpace && ( (index + 1) < valLen) && (isSpace( value[ index + 1 ] ) );

    return ( isLeadingSpace || isTrailingSpace || isNotLastInSequentialSpaces ) ;
```

Reply:
Disagree with this since the triple condition for the last bool is hard to read, and the original comments are clear enough as to what's going on. There's no short circuit evaluation here so all condition are evaluated, unlike the `else if` case, where it only evaluates as many conditions as is needed.

Reply:
I generally think it carries better maintenance value in writing code in an expressive style but maybe in this case the advantage of short-circuit evaluation of `if-else` statements may be important as this function is called for all characters of header pairs.  

@gshvang, please do add the assert check though to the function

Reply:
Yeah I added the assert check.